infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,188 @@
+@Slf4j
+public class AutoCommitService implements ManagementService {
+
+    private static final int COMMIT_BATCH_SIZE = 500;
+    private static final int MAX_COMMIT_RETRY = 8;
+    private static final Duration CONN_RETRY_RATE = Duration.ofMillis(500);
+
+    private final ServerContext serverContext;
+    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
+    private final ScheduledExecutorService autoCommitScheduler;
+
+    // Global log tail fetched at last auto commit cycle, which
+    // would be the commit upper bound in the current cycle.
+    private long lastLogTail = Address.NON_ADDRESS;
+    // Cached committed tail so that we do not need to fetch every cycle.
+    private long committedTail = Address.NON_ADDRESS;
+
+    AutoCommitService(@NonNull ServerContext serverContext,
+                      @NonNull SingletonResource<CorfuRuntime> runtimeSingletonResource) {
+        this.serverContext = serverContext;
+        this.runtimeSingletonResource = runtimeSingletonResource;
+        this.autoCommitScheduler = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setDaemon(true)
+                        .setNameFormat(serverContext.getThreadPrefix() + "AutoCommitService")
+                        .build());
+    }
+
+    CorfuRuntime getCorfuRuntime() {
+        return runtimeSingletonResource.get();
+    }
+
+    /**
+     * Starts the long running service.
+     *
+     * @param interval interval to run the service
+     */
+    @Override
+    public void start(Duration interval) {
+        autoCommitScheduler.scheduleAtFixedRate(
+                () -> LambdaUtils.runSansThrow(this::runAutoCommit),
+                interval.toMillis() / 2,
+                interval.toMillis(),
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    @VisibleForTesting
+    void runAutoCommit() {
+        long lastEpoch = -1;
+
+        for (int i = 1; i <= MAX_COMMIT_RETRY; i++) {
+            try {
+                // Do not perform auto commit if current node is not primary sequencer.
+                if (!isNodePrimarySequencer(updateLayoutAndGet())) {
+                    resetTails();
+                    return;
+                }
+
+                log.debug("runAutoCommit: start committing addresses.");
+
+                // Initialize lastLogTail and committedTail if the first time. We only
+                // commit up to the global tail at the time of last auto commit cycle.
+                if (!Address.isAddress(lastLogTail)) {
+                    initializeTails();
+                    return;
+                }
+
+                // Fetch maximum trim mark from log units and compare with committed tail.
+                // In order not to fetch trim mark for every retry, we only do this when
+                // the epoch changes after last attempt.
+                long currEpoch = getCorfuRuntime().getLayoutView().getLayout().getEpoch();
```

Comment:
I understand that this is an optimization, however, I am pondering if it can result in any issues. For example, what if we have a one node setup, where the epochs change infrequently, but log grows fast, so gets trimmed a lot. Will we by accident committing addresses that have already been trimmed?

Reply:
The if we try to inspect the address brfore the trim mark, log unit server will throw a TrimmedException. Also committing addresses that have trimmed should be safe.